WarehouseManagement.Core/Services/VendorService.cs

```diff
@@ -0,0 +1,232 @@
+        {
+            var vendor = await repository
+                .All<Vendor>()
+                .Where(v => v.Id == id)
+                .Include(v => v.Deliveries)
+                .FirstOrDefaultAsync();
+
+            if (vendor == null)
+            {
+                throw new KeyNotFoundException($"Vendor with ID {id} not found.");
+            }
+
+            if (vendor.Deliveries.Any())
+            {
+                var deliveries = vendor.Deliveries.Select(d => d.SystemNumber).ToList();
+                throw new InvalidOperationException(
+                    $"Vendor can not be deleted because has existing deliveries with system numbers: {string.Join(", ", deliveries)}"
+                );
+            }
+
+            vendor.DeletedByUserId = userId;
+
+            repository.SoftDelete(vendor);
+
+            await repository.SaveChangesAsync();
+        }
+
+        public async Task<string> RestoreAsync(int id)
+        {
+            var vendor = await repository.All<Vendor>().FirstOrDefaultAsync(v => v.Id == id);
+
+            if (vendor == null)
+            {
+                throw new KeyNotFoundException($"Vendor with ID {id} not found.");
+            }
+
+            if (!vendor.IsDeleted)
+            {
+                throw new InvalidOperationException($"Vendor {vendor.Name} is not deleted");
+            }
+
+            if (await this.AnotherVendorWithNameExistAsync(vendor.Id, vendor.Name))
+            {
+                throw new InvalidOperationException(
+                    $"Another vendor with name {vendor.Name} already exist"
+                );
+            }
+
+            if (await this.AnotherVendorWithSystemNumberExistAsync(vendor.Id, vendor.SystemNumber))
+            {
+                throw new InvalidOperationException(
+                    $"Another vendor with system number {vendor.SystemNumber} already exist"
+                );
+            }
+
+            repository.UnDelete(vendor);
+            await repository.SaveChangesAsync();
+
+            return vendor.Name;
+        }
+
+        public async Task<bool> ExistByNameAsync(string name)
+        {
+            return await repository
+                .AllReadOnly<Vendor>()
+                .AnyAsync(v => v.Name.ToLower() == name.ToLower());
+        }
+
+        public async Task<bool> ExistBySystemNumberAsync(string systemNumber)
+        {
+            return await repository
+                .AllReadOnly<Vendor>()
+                .AnyAsync(v => v.SystemNumber.ToLower() == systemNumber.ToLower());
+        }
+
+        public async Task<bool> AnotherVendorWithNameExistAsync(int id, string name)
+        {
+            return await repository
+                .AllReadOnly<Vendor>()
+                .Where(v => v.Id != id)
+                .AnyAsync(v => v.Name.ToLower() == name.ToLower());
+        }
+
+        public async Task<bool> AnotherVendorWithSystemNumberExistAsync(int id, string systemNumber)
```

Comment:
same here as the above question

Reply:
```c#
public async Task<bool> ExistBySystemNumberAsync(string systemNumber)
{
    return await repository
        .AllReadOnly<Vendor>()
        .AnyAsync(v => v.SystemNumber.ToLower() == systemNumber.ToLower());
}

 public async Task<bool> AnotherVendorWithSystemNumberExistAsync(int id, string systemNumber)
 {
     return await repository
         .AllReadOnly<Vendor>()
         .Where(v => v.Id != id)
         .AnyAsync(v => v.SystemNumber.ToLower() == systemNumber.ToLower());
 }
```
same functionality and logic as the above answer. 
`AnotherVendorWithSystemNumberExistAsync` also can be removed and logic implemented in the edit service.

Reply:
maybe we can hide this to the outside (private instead of public) if you think they wont  be needed outside the service?